### ISSUE TITLE
Main Menu Navigation Panel Submenus Contents are Expanded on Keyboard Focus

### ISSUE THREAD

**Issue by Santoz2** [pressbooks/pressbooks#1772]
### Description

When attempting to use a keyboard to interact with the Main Menu navigation panel on the left side of the admin interface, placing focus on any link with a submenu expands the submenu by default. A keyboard user would have to navigate through every item in the submenu to reach the next item.

### Steps to Reproduce

1. Navigate to the My Catalog page in the admin interface
2. Navigate to the Main Menu navigation panel on the left side of the page
3. With a keyboard, use tab, shift + tab, or arrow keys to navigate to a link with a submenu
4. Notice how each submenu is opened by default when focus is placed on the main link, and every item in the submenu must be navigated through to reach the next submenu.

**Expected behavior:** Submenus should be closed unless a user intentionally opens them. As the element that opens a submenu in this case is a link with its own function, it would be best to change the implementation: Possible solutions include adding a separate button to open the submenu, or turning the submenu into children of the main link in a tree view. Examples of keyboard accessible submenus can be seen here: [https://www.w3.org/WAI/tutorials/menus/flyout/](https://www.w3.org/WAI/tutorials/menus/flyout/)

**Actual behavior:** In the Main Menu navigation panel, submenus are opened by default, requiring a keyboard user to navigate through every item in every submenu to reach the end of the list of links.

**Comment by avaaec** [comment]
Additionally all of the menu buttons in the Main Menu navigation region link to the same page as the first item in their sub menu. The link on the menu button should be removed as it is redundant and also a big accessibility issue.

**Comment by SteelWagstaff** [comment]
I think this might be core WordPress behavior and better discussed upstream. To look into further.

**Comment by avaaec** [comment]
This is a UW ATS priority 1 issue.

**Comment by SteelWagstaff** [comment]
@greatislander This ticket seems like it might also be related to upstream WordPress behavior (at least I think I've seen similar behavior for items in the core WordPress dashboard). Do you know anything about flyout menus in WP that would help us file this upstream or implement a better solution for keyboard navigation?

**Comment by greatislander** [comment]
This is core WordPress behaviour and should be filed as an issue on https://core.trac.wordpress.org.

**Comment by SteelWagstaff** [comment]
We brought this issue up in WordPress Slack in the accessibility channel. A core contributor and member of the accessibility team replied:

> The admin menu always worked that way. But, it changes behavior depending on the viewport width. On small screens or if the browser window is not maximized (which sometimes happens to screen reader users without they even have a clue of that) the menu items require activation to reveal the sub menus.

> I wouldn’t say the behavior on large screens is a problem and there’s really no “expected behavior” like the report above claims.

> Imagine the admin menu on desktop like an ordinary unordered list with nested lists: tabbing through the menu all the items can be navigated to. I’d say that’s the simplest possible interaction.
That said, I wouldn’t recommend to even touch the admin menu code as it needs to be backwards compatible with plugins that add their own menu items. Changes to that code will likely break a lot of things. I don’t see great chances to change it outside of a major refactoring that would break backwards compatibility.

> And I wouldn’t say the current behavior is a real problem as the report claims. There’s no established pattern from an accessibility perspective and having the submenus that open while tabbing isn’t an issue, to me.

In response to @Santoz2 suggestion that "The link on the menu button should be removed as it is redundant and also a big accessibility issue." they wrote:

> Yes it is redundant, but it is needed. Claiming it’s a “big issue” sounds to me a bold claim. The menu is keyboard accessible. Keyboard users have to navigate through all the items? Yes. I’d like to understand why this is considered an issue by the reporter.

I followed up and suggested that the problem (as I understood it) was that a user with visual impairment who was navigating the site via keyboard tabbing was forced to tab through every submenu option and that they would have preferred/expected a menu that let users navigating by keyboard choose whether or not to open/navigate submenus. That prompted the following reply: 

> Actually, screen reader users have tons of shortcuts to jump through a list like the one used for the menu so they have way more options for an efficient navigation compared to sighted keyboard users who don’t use a screen reader.

> For this second group of users yes, they have to tab through all the items. Still, I don’t fully understand why this is considered an issue. As a sighted keyboard user I can just hold the tab key pressed and navigate the whole menu in a second. Instead, requiring to activate a menu to display its sub items would slow me down a lot.

> Before assuming there’s an actual problem or pointing to other interaction patterns it would be best to clearly identify the problem, if there’s one. For the record, the pattern with menu items that use a button to reveal their sub items is used on some bundled themes ...

> Aside: I would argue that the example on the W3C tutorial would be hardly usable by speech recognition software users. For example, they would have to guess that the voice command to open the sub menu is `Click show Space Bears submenu` which is unlikely to happen

> I’m guessing that for a blind screen reader user it doesn’t really matter if the sub items open in these so called “fly outs”. All the sub items are announced regardless of their visual presentation. The user is unaware a fly-out is actually displayed on the screen.

> There’s a minor issue regarding the usage of aria-haspopup=true combined with the way the admin menu changes on large and small screens. aria-haspopup=true should only be used when a menu item requires activation to reveal the sub items.

> On large screens, there’s no need for any activation: the sub items are revealed on the go thus aria-haspopup=true should not be used.

> Instead, on small screens the menu changes and the top level items to require activation to reveal the sub items. Only in this case there’s the need to use aria-haspopup=true to inform users there’s a sub menu that appears on activation.

> The changing behavior adds a layer of complexity that is potentially confusing. Personally, I’d rather see a simpler implementation that always works in the same way regardless of the screen size. Maybe there are a couple small improvements that we may want to consider:
> 1. use aria-haspopup=true only for the small screens version: this means aria-haspopup=true should be removed from the markup and added via JavaScript when the menu changes
> 2. explore a way to add keyboard shortcuts for sighted keyboard users, for example to jump through the top level items, with all the caveats custom keyboard shortcuts bring in

A few other screen reader users chimed in -- one of whom stated: 

> As a screen reader user ... this doesn't bother me at all. I have been in discussions with sighted keyboard users before, some of who preferred to have to open sub menus, others not, for exactly the reasons stated above.
